Add screen-related attributes to mouse input events

Sauermann opened the pull request:
Add screen velocity and screen relative position to `InputEventMouseMotion` and `InputEventScreenDrag`.
Based on discussion in RC.

- *Production edit: This closes https://github.com/godotengine/godot/issues/34805.*

Sauermann committed:
Add screen-related attributes to mouse input events

Calinou commented:
Does this PR address https://github.com/godotengine/godot/issues/30950?

Sauermann commented:
> Does this PR address #30950?

No, because it doesn't target 3.x. In 4.x #30950 should already be resolved.

groud reviewed (approved):
This was discussed in an Input PR meeting, and was approved!
Thanks.

Sauermann commented on core/input/input_event.h:
```diff
@@ -415,9 +425,15 @@ class InputEventScreenDrag : public InputEventFromWindow {
 	void set_relative(const Vector2 &p_relative);
 	Vector2 get_relative() const;
 
+	void set_relative_screen_position(const Vector2 &p_velocity);
```
Thanks for the two corrections. Done.

akien-mga commented:
> How should those fields be filled by platforms that can't query such data? Wayland is mostly window-centric and the mouse position can only be queried in window-space when it's on a surface.

Poke @Sauermann

Sauermann commented:
This point was discussed on rocket chat. See https://chat.godotengine.org/channel/platforms?msg=777ZLDf26xiuswuou
If I remember correctly, On wayland the relative attributes can also be calculated.

akien-mga commented:
Sounds good. I guess we should evaluate whether we want to merge #86180 first, or whether the needed changes for this PR could also be added to #86180 (as you prefer @Riteo).

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/63d6bda8e95ac992da74e84b2f3be62f3d85190b), it works as expected on Linux in all stretch modes.

This seems to be exactly what I want with [`relative_unscaled` and `velocity_unscaled` coordinates](https://github.com/Calinou/godot/tree/inputevent-add-relative-velocity-unscaled), albeit with a different name. (reduz also proposed `global_relative` and `global_velocity` terms for this feature, but I can't find the exact issue/PR.)

In terms of naming, I still prefer `relative_unscaled` and `velocity_unscaled` (as I think they're the most likely terms to be searched by users), even if they are less technically correct.

We should make sure to update documentation and demos to use these new properties if we merge this PR.

**Testing project:** [test_unscaled_relative_input_2.zip](https://github.com/godotengine/godot/files/14167972/test_unscaled_relative_input_2.zip)

Calinou commented on doc/classes/InputEventMouseMotion.xml:
```diff
@@ -24,6 +24,13 @@
 			The mouse position relative to the previous position (position at the last frame).
 			[b]Note:[/b] Since [InputEventMouseMotion] is only emitted when the mouse moves, the last event won't have a relative position of [code]Vector2(0, 0)[/code] when the user stops moving the mouse.
```
My branch has additional docs, so I suggest porting them over to describe why you should use one property over the other:

```suggestion
			[b]Note:[/b] Since [InputEventMouseMotion] is only emitted when the mouse moves, the last event won't have a relative position of [code]Vector2(0, 0)[/code] when the user stops moving the mouse.
			[b]Note:[/b] [member relative] is automatically scaled according to the content scale factor, which is defined by the project's stretch mode settings. This means mouse sensitivity will appear different depending on resolution when using [member relative] in a script that handles mouse aiming with the [constant Input.MOUSE_MODE_CAPTURED] mouse mode. To avoid this, use [member screen_relative] instead.
```

Calinou commented on doc/classes/InputEventMouseMotion.xml:
```diff
@@ -24,6 +24,13 @@
 			[b]Note:[/b] Since [InputEventMouseMotion] is only emitted when the mouse moves, the last event won't have a relative position of [code]Vector2(0, 0)[/code] when the user stops moving the mouse.
 		</member>
+		<member name="screen_relative" type="Vector2" setter="set_screen_relative" getter="get_screen_relative" default="Vector2(0, 0)">
+			The mouse position relative to the previous position in the coordinate system of the screen (position at the last frame).
```
```suggestion
			The mouse position relative to the previous position in the coordinate system of the screen (position at the last frame). This coordinate is [i]not[/i] scaled according to the content scale factor or calls to [method InputEvent.xformed_by]. This should be preferred over [member relative] for mouse aiming when using the [constant Input.MOUSE_MODE_CAPTURED] mouse mode, regardless of the project's stretch mode.
```

Calinou commented on doc/classes/InputEventMouseMotion.xml:
```diff
@@ -24,6 +24,13 @@
+			[b]Note:[/b] Since [InputEventMouseMotion] is only emitted when the mouse moves, the last event won't have a relative position of [code]Vector2(0, 0)[/code] when the user stops moving the mouse.
+		</member>
+		<member name="screen_velocity" type="Vector2" setter="set_screen_velocity" getter="get_screen_velocity" default="Vector2(0, 0)">
+			The mouse velocity in pixels per second in screen coordinates.
```
```suggestion
			The mouse velocity in pixels per second in screen coordinates. This velocity is [i]not[/i] scaled according to the content scale factor or calls to [method InputEvent.xformed_by]. This should be preferred over [member velocity] for mouse aiming when using the [constant Input.MOUSE_MODE_CAPTURED] mouse mode, regardless of the project's stretch mode.
```

Also, could you add this note to the `velocity` description?

```
[b]Note:[/b] [member velocity] is automatically scaled according to the content scale factor, which is defined by the project's stretch mode settings. This means mouse sensitivity will appear different depending on resolution when using [member velocity] in a script that handles mouse aiming with the [constant Input.MOUSE_MODE_CAPTURED] mouse mode. To avoid this, use [member screen_velocity] instead.
```

Calinou commented on doc/classes/InputEventScreenDrag.xml:
```diff
@@ -25,6 +25,12 @@
 			The drag position relative to the previous position (position at the last frame).
 		</member>
+		<member name="screen_relative" type="Vector2" setter="set_screen_relative" getter="get_screen_relative" default="Vector2(0, 0)">
+			The drag position relative to the previous position in screen coordinates (position at the last frame).
```
```suggestion
			The drag position relative to the previous position in screen coordinates (position at the last frame). This position is [i]not[/i] scaled according to the content scale factor or calls to [method InputEvent.xformed_by]. This should be preferred over [member relative] for touch aiming regardless of the project's stretch mode.
```

Calinou commented on doc/classes/InputEventScreenDrag.xml:
```diff
@@ -25,6 +25,12 @@
+			The drag position relative to the previous position in screen coordinates (position at the last frame).
+		</member>
+		<member name="screen_velocity" type="Vector2" setter="set_screen_velocity" getter="get_screen_velocity" default="Vector2(0, 0)">
+			The drag velocity in pixels per second in screen coordinates.
```
```suggestion
			The drag velocity in pixels per second in screen coordinates. This velocity is [i]not[/i] scaled according to the content scale factor or calls to [method InputEvent.xformed_by]. This should be preferred over [member velocity] for touch aiming regardless of the project's stretch mode.
```

kitbdev commented:
If `unscaled` is more likely to be searched, could we add that word into the documentation?

Calinou commented:
> If `unscaled` is more likely to be searched, could we add that word into the documentation instead of changing the name?

https://github.com/godotengine/godot/pull/78990 would allow adding a search alias :slightly_smiling_face:

Sauermann commented:
I have updated the PR:
- Rebased on master after #86180 and added changes to Wayland. (Untested)
- Included multiple documentation additions by Calinou. Please let me know if I should mention you as a co-author.
- Included word "unscaled" into descriptions.

Sauermann commented:
In the displayserver-context both are identical.

Scaling happens in `Viewport::push_input`:
https://github.com/godotengine/godot/blob/d3352813ea44447bfbf135efdec23acc4d1d3f89/scene/main/viewport.cpp#L3338-L3343

Sauermann commented:
@Riteo partially these values have been in the engine for a very long time. After #37317 perhaps it makes sense to reevaluate these aspects.

akien-mga commented:
Thanks!
